File: Assets/InsightARWorld/InsightARExporter/Utility/BuiltinJSScripts/Location/CalculateArrowPosition.js
function CalculateArrowPosition(gameObject)
{
    this.gameObject = gameObject;
    this.transform = gameObject.transform;
}
//Write prototype function here
CalculateArrowPosition.prototype = Object.assign(Object.create(Object.prototype), {
    // Start is called before the first frame update
    // Start: function(){

    // },
    // // Update is called once per frame
    // Update: function()
    // {

    // },
    // calculate rotation
    CalculateRotation: function(camera,stuffPosition)
    {
        var angle = this.CalculateAngle(camera,stuffPosition);
        return Insight.Quaternion.AngleAxis(angle,new Insight.Vector3(0, 0, 1));
    },
    // CalculatePosition
    CalculatePosition: function(camera,trans,stuffPosition,screenHeight,screenWidth,marginH,marginV)
    {
        var angle = this.CalculateAngle(camera,stuffPosition);
        var cosA = Math.cos(Fw_Math_Rad(1) * angle);
        var sinA = Math.sin(Fw_Math_Rad(1) * angle);
        
        var posZ = trans.gameObject.getComponent("RectTransform").anchoredPosition3D.z;
        return new Insight.Vector3((screenWidth - marginH)/2.0*cosA,(screenHeight - marginV)/2.0 * sinA,posZ);
    },
    // calculate angle
    CalculateAngle: function(camera,stuffPosition)
    {
        var camTrans = camera.transform;
        var stuffOffset = new Insight.Vector3();
        stuffOffset.copy(stuffPosition).sub(camTrans.position);

        var camRight = camTrans.right;
        var camUp = camTrans.up;
        var camForward = camTrans.forward;
        
        var cosRight = camRight.dot(stuffOffset)/stuffOffset.magnitude;//Insight.Vector3.Dot(camRight,stuffOffset)/stuffOffset.magnitude;
        var cosUp = camUp.dot(stuffOffset)/stuffOffset.magnitude;//Insight.Vector3.Dot(camUp,stuffOffset)/stuffOffset.magnitude;
        
        var length = Math.sqrt(cosRight * cosRight + cosUp * cosUp);
        
        if(length < 0.001)
        {
            return 0.0;
        }
        
        var cosA = cosRight/length;
        var sinA = cosUp /length;
        
        var angle = 0;
        
        if(sinA >=0 && cosA >= 0)
        {
            angle = Fw_Math_Deg(1)*Math.asin(sinA);
        }
        else if(sinA>=0 && cosA<0)
        {
            angle = Fw_Math_Deg(1) * Math.acos(cosA);
        }
        else if(sinA<0 && cosA <0)
        {
            angle = 180 - Fw_Math_Deg(1) * Math.asin(sinA);
        }
        else
        {
            angle = Fw_Math_Deg(1) * Math.asin(sinA);
        }
        return angle;
    }
});

//Return the script module
CalculateArrowPosition

File: Assets/InsightARWorld/InsightARExporter/Preview/Scripts/Common/Component/CalculateArrowPosition.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace Dongjian.LargeScale
{
    /// <summary>
    /// 计算UI箭头位置
    /// </summary>
    public class CalculateArrowPosition
    {
        #region utility
        /// <summary>
        /// 计算图标rotation
        /// </summary>
        /// <returns></returns>
        public static Quaternion CalculateRotation(Camera camera, Vector3 stuffPosition)
        {

            float angle = CalculateAngle(camera, stuffPosition);
            return Quaternion.AngleAxis(angle, new Vector3(0, 0, 1));

        }

        /// <summary>
        /// 计算图标anchorposition
        /// </summary>
        /// <param name="camera"></param>
        /// <param name="trans"></param>
        /// <returns></returns>
        public static Vector3 CalculateAnchorPosition(Camera camera, RectTransform rectTrans,Vector3 stuffPosition,
            int screenHeight = 1920,int screenWidth = 1080,int marginHorizontal = 160,int marginVertical = 320)
        {
            float angle = CalculateAngle(camera, stuffPosition);
            float cosA = Mathf.Cos(Mathf.Deg2Rad * angle);
            float sinA = Mathf.Sin(Mathf.Deg2Rad * angle);

            float positionZ = rectTrans.anchoredPosition3D.z;
            return new Vector3((screenWidth - marginHorizontal) / 2.0f * cosA, (screenHeight - marginVertical) / 2.0f * sinA, positionZ);
        }


        /// <summary>
        /// 计算图标的角度
        /// </summary>
        /// <param name="camera"></param>
        /// <param name="trans"></param>
        /// <returns></returns>
        private static float CalculateAngle(Camera camera, Vector3 stuffPosition)
        {
            Transform camTrans = camera.transform;
            Vector3 stuffOffset = stuffPosition - camTrans.position;
            Vector3 camRight = camTrans.right;
            Vector3 camUp = camTrans.up;
            Vector3 camForward = camTrans.forward;

            float cosRight = Vector3.Dot(camRight, stuffOffset) / stuffOffset.magnitude; //在相机right轴投影

            float cosUp = Vector3.Dot(camUp, stuffOffset) / stuffOffset.magnitude; //在相机up轴投影


            float length = Mathf.Sqrt(cosRight * cosRight + cosUp * cosUp);
            if (length < 0.0001f)
                return 0.0f;

            // 在right - O - up平面，计算投影角度
            float cosA = cosRight / length;
            float sinA = cosUp / length;


            float angle = 0;

            if (sinA >= 0 && cosA >= 0)  //第一象限
                angle = Mathf.Rad2Deg * Mathf.Asin(sinA);
            else if (sinA >= 0 && cosA < 0)    //第二象限
                angle = Mathf.Rad2Deg * Mathf.Acos(cosA);
            else if (sinA < 0 && cosA < 0)    //第三象限
                angle = 180 - Mathf.Rad2Deg * Mathf.Asin(sinA);
            else  //第四象限
                angle = Mathf.Rad2Deg * Mathf.Asin(sinA);

            return angle;

        }
        #endregion

    }
}
